shopify_function_macro/src/lib.rs

```diff
@@ -496,35 +508,76 @@ impl CodeGenerator for ShopifyFunctionCodeGenerator {
     ) -> Vec<syn::ItemImpl> {
         let name_ident = names::type_ident(input_object_type_definition.name());
 
+        // Conditionally serialize fields based on GraphQL schema nullability
+        // Nullable fields (Option<T>) are only serialized if Some(_)
+        // Required fields are always serialized
+
         let field_statements: Vec<syn::Stmt> = input_object_type_definition
             .input_field_definitions()
             .iter()
             .flat_map(|ivd| {
                 let field_name_ident = names::field_ident(ivd.name());
                 let field_name_lit_str = syn::LitStr::new(ivd.name(), Span::mixed_site());
 
-                vec![
+                // Check if this field is nullable in the GraphQL schema
+                if is_input_field_nullable(ivd) {
+                    // For nullable fields, only serialize if Some(_)
+                    vec![parse_quote! {
+                        if let ::std::option::Option::Some(ref value) = self.#field_name_ident {
+                            context.write_utf8_str(#field_name_lit_str)?;
+                            value.serialize(context)?;
+                        }
+                    }]
+                } else {
+                    // For required fields, always serialize
+                    vec![
+                        parse_quote! {
+                            context.write_utf8_str(#field_name_lit_str)?;
+                        },
+                        parse_quote! {
+                            self.#field_name_ident.serialize(context)?;
+                        },
+                    ]
+                }
+            })
+            .collect();
+
+        // Generate field counting statements for dynamic field count calculation
+        let field_count_statements: Vec<syn::Stmt> = input_object_type_definition
+            .input_field_definitions()
+            .iter()
+            .map(|ivd| {
+                let field_name_ident = names::field_ident(ivd.name());
+
+                if is_input_field_nullable(ivd) {
+                    // For nullable fields, count only if Some(_)
                     parse_quote! {
-                        context.write_utf8_str(#field_name_lit_str)?;
-                    },
+                        if let ::std::option::Option::Some(_) = self.#field_name_ident {
+                            field_count += 1;
+                        }
+                    }
+                } else {
+                    // For required fields, always count
                     parse_quote! {
-                        self.#field_name_ident.serialize(context)?;
-                    },
-                ]
+                        field_count += 1;
+                    }
+                }
             })
             .collect();
```

Comment:
Can we just consolidate this into one line where we pre-compute the number of non-nullable fields and add the rest based on value, e.g. we might end up with:
```rust
let field_count: usize = 5 + self.foo.is_some().into() + self.bar.is_some().into();
```